src/treesitter.rs

```diff
@@ -21,3 +30,305 @@ impl From<&'_ Node<'_>> for Match {
+    tree: Cow<'tree, Tree>,
+    capture_index: u32,
+    #[borrows(text, mut query_cursor, query, tree)]
+    #[covariant]
+    captures_iterator: QueryCaptures<'this, 'this, 'this, RopeOrSlice<'this>>,
+    #[borrows(tree)]
+    #[covariant]
+    next_capture: Option<CaptureInfo<'this>>,
+}
+
+pub fn get_captures<'a, 'text, 'tree>(
+    language: Language,
+    // text: impl TextProvider<'a> + Parseable,
+    text: impl Into<RopeOrSlice<'text>>,
+    query: &'a Query,
+    capture_index: u32,
+    filter: Option<&'a Filterer>,
+    tree: Option<&'tree Tree>,
+) -> Captures<'a, 'text, 'tree> {
+    let text = text.into();
+    let query_cursor = QueryCursor::new();
+    let tree: Cow<'tree, Tree> = tree.map_or_else(
+        || Cow::Owned(text.parse(&mut get_parser(language), None).unwrap()),
+        Cow::Borrowed,
+    );
+    Captures::new(
+        text,
+        query_cursor,
+        query,
+        filter,
+        tree,
+        capture_index,
+        |text, query_cursor, query, tree| query_cursor.captures(query, tree.root_node(), *text),
+        |_| None,
+    )
+}
+
+impl<'a, 'text, 'tree> StreamingIterator for Captures<'a, 'text, 'tree> {
+    type Item = CaptureInfo<'tree>;
+
+    fn advance(&mut self) {
+        self.with_mut(|all_fields| {
+            for (match_, index_into_query_match_captures) in all_fields.captures_iterator.by_ref() {
+                let this_capture = &match_.captures[index_into_query_match_captures];
+                if this_capture.index != *all_fields.capture_index {
+                    continue;
+                }
+                let single_captured_node = this_capture.node;
+                if all_fields
+                    .filter
+                    .as_ref()
+                    .map_or(true, |filter| filter.call(&single_captured_node))
+                {
+                    *all_fields.next_capture = Some(CaptureInfo {
+                        node: single_captured_node,
+                        pattern_index: match_.pattern_index,
+                    });
+                    return;
+                }
+            }
+            *all_fields.next_capture = None;
+        });
+    }
+
+    fn get<'this>(&'this self) -> Option<&'this Self::Item> {
+        let next_capture = self.borrow_next_capture();
+        // SAFETY: I think this is ok as long as CaptureInfo isn't
+        // Copy/Clone?
+        // Since at that point there's no way for the "inner"
+        // CaptureInfo's contents to "outlive" the returned reference?
+        // Did this because otherwise was running into not being able
+        // to express that the "real" Item type for this trait (I think)
+        // should be CaptureInfo<'this>, not CaptureInfo<'a>
+        let next_capture: &'this Option<CaptureInfo<'tree>> =
+            unsafe { mem::transmute(next_capture) };
+        next_capture.as_ref()
+    }
+}
+
+#[self_referencing]
+pub struct CapturesForEnclosingNode<'a, 'text: 'a, 'tree: 'a> {
```

Comment:
Same thing, just for the variation (as used by `tree-sitter-lint`) where you want to provide an explicit "enclosing node" for the query vs it defaulting to querying against the entire file contents associated with the `Tree`